cirq/circuits/circuit.py

```diff
@@ -1314,6 +1324,49 @@ def _draw_moment_in_diagram(moment: Moment,
         if exponent is not None:
             out_diagram.write(x, y2, '^' + exponent)
 
+    # Group together columns belonging to the same Moment.
+    if moment.operations and x > x0:
+        moment_groups.append([x0, x])
+
+
+def _draw_moment_groups_in_diagram(moment_groups: List[Tuple[int, int]],
+                                   use_unicode_characters: bool,
+                                   out_diagram: TextDiagramDrawer,
+                                   transpose: bool):
+    out_diagram.shift_vertically(0, 1)
+    h = out_diagram.height()
+
+    left_end = '├' if use_unicode_characters else '|'
+    top_end = '┬' if use_unicode_characters else '-'
+    right_end = '┤' if use_unicode_characters else '|'
+    bottom_end = '┴' if use_unicode_characters else '-'
+
+    # Insert columns starting from the back since the insertion
+    # affects subsequent indices.
+    for (x1, x2) in reversed(moment_groups):
```

Comment:
The efficiency of this loop can be improved by doing all the padding at once. I'll leave it this way for now as it's easier to reason about. I can make the improvement later.